Initial Fibaro HC Climate support

pbalogh77 opened the pull request:
## Description:
Initial support for climate platform for the Fibaro hub integration.

**Related issue (if applicable):** fixes #19359

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8748

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

pbalogh77 committed:
Initial version of climate

pbalogh77 committed:
initial commit of climate device support

pbalogh77 committed:
Fixed opmode and fanmode

pbalogh77 committed:
Cleanup

pbalogh77 committed:
meh

pbalogh77 committed:
added back all other components

Oops

pbalogh77 committed:
wider support for thermostats

Added one more identifier for thermostats to broaden compatibility

pbalogh77 committed:
Added even more climate types

pbalogh77 committed:
Reworked detection mechanism

Better support for combined devices

pbalogh77 committed:
Added additional modes

pbalogh77 committed:
force visibility on climate

pbalogh77 committed:
Changed logging of device data

pbalogh77 committed:
Improved operatingmode support

Improved operatingmode support

elbogi commented:
Any progress on this PR, or anything that the community could do to help this get into the next RC?

pbalogh77 committed:
Updated logic for opmode/fanmode list creation

Implemented a universal mapping logic for opmode and fanmode, to make it more widely compatible
Improved mapping of Fibaro FGT devices

pbalogh77 committed:
Lint fixes

pbalogh77 commented:
Docs PR: https://github.com/home-assistant/home-assistant.io/pull/8748

pbalogh77 committed:
bump

MartinHjelmare commented on homeassistant/components/climate/fibaro.py:
```diff
@@ -0,0 +1,272 @@
+            return False
+        return True
+
+    def update(self):
```
Remove this since we don't use it.

MartinHjelmare commented on homeassistant/components/fibaro/__init__.py:
```diff
@@ -174,6 +169,18 @@ def register(self, device_id, callback):
+    def get_children(self, device_id):
+        """Get a list of child devices."""
+        result = []
+        for device in self._device_map.values():
```
```py
return [
    device for device in self._device_map.values()
    if device.parentId == device_id]
```

pbalogh77 replied:
Thanks! Much cleaner!

MartinHjelmare reviewed (changes requested):
Please `git mv` the fibaro climate platform to the fibaro component package. That's our new way of structuring integrations.
```
homeassistant/components/fibaro/climate.py
```

pbalogh77 commented on homeassistant/components/climate/fibaro.py:
```diff
@@ -0,0 +1,272 @@
+SPEED_LOW = 'low'
+SPEED_MEDIUM = 'medium'
+SPEED_HIGH = 'high'
+STATE_AUXILIARY = 'auxiliary'
```
But these devices are managed by a Fibaro hub, so if they are in a state that is not support by HA, I should still understand that, I think. The opmode states are coming from the zwave specifications. So I'm not sure what to do in case HA does not support an opmode state that the device is in.

MartinHjelmare replied:
We can only map as best as possible to the operation modes that home assistant supports. If those don't make sense we have to leave the other modes out.

It's ok to add device state attributes that show extra information related to the state.

pbalogh77 replied:
Is there a list somewhere of supported modes by HA?

pbalogh77 commented on homeassistant/components/climate/fibaro.py:
```diff
@@ -0,0 +1,272 @@
+                       properties.operatingMode)
+        else:
+            mode = int(self._opmode_device.fibaro_device.properties.mode)
+        ret = self._opmode_map.get(mode, "")
```
So like this, right?

        return self._op_mode_to_state.get(mode)

MartinHjelmare replied:
Yes! :+1:

pbalogh77 committed:
Fixes based on code review

pbalogh77 committed:
Fixes

pbalogh77 committed:
Moved to fibaro folder

pbalogh77 committed:
lint inspired cosmetic changes

MartinHjelmare commented on homeassistant/components/fibaro/climate.py:
```diff
@@ -0,0 +1,276 @@
+    1: STATE_HEAT,
+    2: STATE_COOL,
+    3: STATE_AUTO,
+    4: STATE_AUXILIARY,
```
Only operation states present in the base climate component are allowed as operation modes.

Please limit the home assistant modes to those states.

pbalogh77 replied:
Done

pbalogh77 committed:
Mapped all operating modes to existing HA ones

Mapped all operating modes to existing HA ones

pbalogh77 committed:
Improved compatibility with Heatit thermostats

Thanks to astrandb for testing, debugging and fixing my code

MartinHjelmare commented on homeassistant/components/fibaro/climate.py:
```diff
@@ -0,0 +1,288 @@
+FAN_LEFT_RIGHT = 'left_right'
+FAN_UP_DOWN = 'up_down'
+FAN_QUIET = 'quiet'
+STATE_FURNACE = 'furnace'
```
This one is still left.

MartinHjelmare commented on homeassistant/components/fibaro/climate.py:
```diff
@@ -0,0 +1,288 @@
+                      if self._target_temp_device else "None")
+        _LOGGER.debug("- _op_mode_device %s", self._op_mode_device.ha_id
+                      if self._op_mode_device else "None")
+        _LOGGER.debug("- _fan_mode_device %s", self._fan_mode_device.ha_id
```
Would be nice to combine these logging statements into one.

pbalogh77 committed:
Changes based on code review

Changes based on code review

MartinHjelmare commented on homeassistant/components/fibaro/climate.py:
```diff
@@ -0,0 +1,291 @@
+            if "setThermostatSetpoint" in target.fibaro_device.actions:
+                target.action("setThermostatSetpoint",
+                              self._op_state_to_mode[self.current_operation],
+                              kwargs.get(ATTR_TEMPERATURE))
```
Use `temperature`.

MartinHjelmare commented on homeassistant/components/fibaro/climate.py:
```diff
@@ -0,0 +1,291 @@
+                              kwargs.get(ATTR_TEMPERATURE))
+            else:
+                target.action("setTargetLevel",
+                              kwargs.get(ATTR_TEMPERATURE))
```
Same as above.

MartinHjelmare reviewed (approved):
Can be merged when last comments are addressed and build passes.

pbalogh77 committed:
more fixes based on more code review

more fixes based on more code review
